engine/buildkit/ref.go

```diff
@@ -348,6 +356,101 @@ func wrapError(ctx context.Context, baseErr error, sessionID string) error {
 	}
 }
 
+func debugContainer(ctx context.Context, execOp *bksolverpb.ExecOp, execErr *llberror.ExecError, client *Client) error {
+	if !client.Opts.Interactive {
+		return nil
+	}
+
+	// If this is the (internal) exec of the module itself, we don't want to spawn a terminal.
+	for _, env := range execOp.Meta.Env {
+		if strings.HasPrefix(env, "_DAGGER_CALL_DIGEST=") {
```

Comment:
This part was a hack to unblock proving the concept. I can live with it (just slightly fragile to future code changes) but ideally we'll be able instead get the `ExecutionMetadata` struct associated with the exec here (https://github.com/sipsma/dagger/blob/66e98c39ef4ed0aa17bbb2c8b47e48b10620bdaa/engine/buildkit/executor.go#L43-L43)

Then we can plumb whatever we want here, which should also help with providing more debug information on what failed container they are even dropping into, customizing to something besids /bin/sh.

Unfortunately on a first check it looks like buildkit is not quite giving us the field we need to extract that from. I'm double checking now and if it's indeed missing will send an upstream PR to support it provided it's as simple as I'm imagining right now.

Not a blocker, after that I'll finish reviewing, just an FYI.

Reply:
Sent out that upstream PR here https://github.com/moby/buildkit/pull/5108

Reply:
@aluzzardi after a rebase on main you should be able to do this:
```diff
diff --git a/engine/buildkit/ref.go b/engine/buildkit/ref.go
index e42f9c591..6542b604f 100644
--- a/engine/buildkit/ref.go
+++ b/engine/buildkit/ref.go
@@ -297,6 +297,14 @@ func wrapError(ctx context.Context, baseErr error, client *Client) error {
                return baseErr
        }
 
+       execMd, ok, err := ExecutionMetadataFromDescription(opErr.Description)
+       if err != nil {
+               // ...
+       }
+       if !ok {
+               // ... this is a valid case for containers created by buildkit internals like the dockerfile frontend
+       }
+
        // This was an exec error, we will retrieve the exec's output and include
        // it in the error message
 ```

Then we can decide whether to open the debug terminal based on that execution metadata (and use anything else or add more fields if useful for customizing the debug container; the only requirement for ExecutionMetadata is that it's json serializable).

That metadata gets initialized here: https://github.com/sipsma/dagger/blob/66e98c39ef4ed0aa17bbb2c8b47e48b10620bdaa/core/container_exec.go#L72-L72
* With some additional input for module functions specifically here: https://github.com/sipsma/dagger/blob/66e98c39ef4ed0aa17bbb2c8b47e48b10620bdaa/core/modfunc.go#L180-L180

So I guess we can decide whether to drop into the shell by default and have a field there for disabling it, or the reverse. I suppose doing it by default and just explicitly disabling for module function containers sounds better to me at first thought, but either could make sense.

Reply:
🙏 

Tried something (last commit) with terminal by default (`Internal` flag on `ExecutionMetadata`), PTAL

This is pretty sweet -- I was thinking we could add `call.ID` or similar in order to be able to print the execution path (what `Terminal()` does using `idtui.DumpID`), but I didn't see an obvious path in getting that

Reply:
Oh I actually am already adding the call ID to ExecutionMetadata in a separate PR (https://github.com/dagger/dagger/pull/7804/files#diff-6e7d0c02c8bd065fb312e4109724ee002cb041cd44c4c6e97abe267ce31cedf4R184), you can get the value from `dagql.CurrentID(ctx)`.

I think you'll also need these lines I added there so the ID json serializes (feel free to just copy-pasta): https://github.com/dagger/dagger/blob/ccdea050df7fbe7fb4237a308bdcc7016ee7337c/dagql/call/id.go#L271-L285